Added dt support for `xtc` and `trr` trajectories.

pbuslaev opened the pull request:
<!-- 
Insert issue number that this PR fixes (if any) just after 'Fixes #'.
If this PR does not fix an existing issue, consider opening one or remove 'Fixes #' from the PR description.
-->
Fixes #4905

Changes made in this Pull Request:
<!-- Describe the changes that this PR makes. If applicable, use the following bullet list. --> 
- I added basic dt support for XTC and XDR Reader/Writer

## PR Checklist
<!-- Please use the following checklist to ensure the PR is ready to be reviewed/merged. -->
 - [x] Issue raised/referenced?
 - [x] Tests updated/added?
 - [x] Documentation updated/added?
 - [x] `package/CHANGELOG` file updated?
 - [x] Is your name in `package/AUTHORS`? (If it is not, add it!)

## Developers Certificate of Origin
<!-- 
In order for this PR to be merged you **must certify that you are able to submit your code to be included in MDAnalysis**.
-->

I certify that I can submit this code contribution as described in the [**Developer Certificate of Origin**](https://developercertificate.org/), under the MDAnalysis [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE).


<!-- readthedocs-preview mdanalysis start -->
----
📚 Documentation preview 📚: https://mdanalysis--4908.org.readthedocs.build/en/4908/

<!-- readthedocs-preview mdanalysis end -->

pbuslaev committed:
Updated xdr and xtc reader to set dt

pbuslaev committed:
Added dt option to xtc writer

pbuslaev committed:
Fixed typo

pbuslaev committed:
Fixed another typo

pbuslaev committed:
Added dt support for trr

pbuslaev committed:
Fixed style

pbuslaev committed:
Fixed style. Added data to Authors and changelog.

pbuslaev committed:
Updated tests

pbuslaev committed:
fixed style in test

pbuslaev committed:
Added missed fixture decoratorfor xdr test

pbuslaev committed:
Fixed failing terst

pbuslaev committed:
Fixing failing tests, by referring to correct universe

pbuslaev committed:
Fixed logics in xdr writers and readers

pbuslaev committed:
Fixed logics to pass failed tests

pbuslaev committed:
Fixed style

pbuslaev committed:
Fixed logics for the reader

orbeckst commented on package/MDAnalysis/coordinates/XDR.py:
```diff
@@ -148,6 +149,8 @@ def __init__(
         refresh_offsets : bool (optional)
             force refresh of offsets
+        dt : float (optional)
+            timestep in MDAnalysis units to load trajectory with
```
Clearly document what happens here.

State that None reads the time as stored in the trajectory. Choosing dt overrides the time that is stored in the trajectory and assigns `frame * dt` as time.

orbeckst commented on package/MDAnalysis/coordinates/TRR.py:
```diff
@@ -113,7 +113,12 @@ def _write_next_frame(self, ag):
-        time = ts.time
+        if self._dt is None:
+            time = ts.time
+        elif self._dt == ts.dt:
+            time = ts.time
```
Delete this case: this leads to weird behavior when the initial frame is at 0 but the time is not. In this case, setting dt will behave differently from what the user would expect, namely that `time = frame * dt`. Make sure that the code exactly implements what you document.

pbuslaev replied:
My thinking here, was that if dt is set, and it is exactly equal to one of the trajectory, I don't want to touch the time in the frame.

orbeckst replied:
I understand but that's not how it's documented. If the user does not want to change anything then they should simply not set `dt`. Less magic is better.

pbuslaev replied:
Ok, I now remember why, I had to add this. Without dt option, tests were breaking.

For instance this [test](https://github.com/pbuslaev/mdanalysis/blob/f4f5e7d3532e77dcde73f4241ec4ea3f90a10dbe/testsuite/MDAnalysisTests/coordinates/test_dcd.py#L300) starts failing, because `dt` is set for the writer. So there are two choices, as I see it:

- Leave this option and document it
- Modify failing tests
- (?) maybe there is some more complex logics that can be implemented

@orbeckst I can take either approach, but I think it is up to you to decide what is better.

orbeckst commented on package/MDAnalysis/coordinates/XDR.py:
```diff
@@ -327,13 +335,16 @@ def __init__(self, filename, n_atoms, convert_units=True, **kwargs):
         convert_units : bool (optional)
             convert from MDAnalysis units to format specific units
+        dt : float (optional)
+            timestep in MDAnalysis units to write trajectory with
```
document exactly what dt does and what the default means

orbeckst commented on package/MDAnalysis/coordinates/XTC.py:
```diff
@@ -102,8 +104,13 @@ def _write_next_frame(self, ag):
+        if self._dt is None:
+            time = ts.time
+        elif self._dt == ts.dt:
+            time = ts.time
```
as in TRR – remove this case because it leads to ambiguous behavior

orbeckst commented on package/MDAnalysis/coordinates/XTC.py:
```diff
@@ -102,8 +104,13 @@ def _write_next_frame(self, ag):
+            time = ts.time
+        else:
+            time = self._dt * ts.frame
+        step = ts.data.get('step', ts.frame)
```
good fix

orbeckst commented on package/MDAnalysis/coordinates/TRR.py:
```diff
@@ -176,7 +181,11 @@ def _read_next_timestep(self, ts=None):
         """convert a trr-frame to a mda TimeStep"""
-        ts.time = frame.time
+        dt = self._kwargs["dt"]
+        if dt is not None:
```
start with the default `dt is None`

orbeckst commented on package/MDAnalysis/coordinates/XTC.py:
```diff
@@ -64,10 +64,12 @@ def __init__(self, filename, n_atoms, convert_units=True,
         convert_units : bool (optional)
             convert into MDAnalysis units
+        dt : float (optional)
+            timestep in MDAnalysis units to load trajectory with
```
document exact behavior, emphasize that `None` is the default that reads time from trajectory and that `dt` as a float is for special cases when one wants to set a different time step from what is stored

orbeckst commented on package/MDAnalysis/coordinates/XTC.py:
```diff
@@ -64,10 +64,12 @@ def __init__(self, filename, n_atoms, convert_units=True,
             set precision of saved trjactory to this number of decimal places.
         """
-        super(XTCWriter, self).__init__(filename, n_atoms, convert_units,
+        super(XTCWriter, self).__init__(filename, n_atoms, convert_units, dt,
```
use kwargs for the super call (sorry that convert_units was not done)
```suggestion
        super(XTCWriter, self).__init__(filename, n_atoms, convert_units=convert_units, dt=dt,
```
(you probably need to line-break ... actually, just run through black)

orbeckst commented on package/MDAnalysis/coordinates/XTC.py:
```diff
@@ -164,7 +171,11 @@ def _read_next_timestep(self, ts=None):
         ts.frame = self._frame
-        ts.time = frame.time
+        dt = self._kwargs["dt"]
+        if dt is not None:
```
Move the default behavior first in if block

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_xdr.py:
```diff
@@ -245,6 +278,31 @@ def test_Writer(self, tmpdir):
+    def test_Writer_with_dt_setting(self, tmpdir):
+        universe = mda.Universe(GRO, self.filename, convert_units=True)
+        ext = os.path.splitext(self.filename)[1]
+        outfile = str(tmpdir.join("/xdr-reader-test" + ext))
```
Do you have to convert to string?  Is Writer not able to use the Path instance?

orbeckst replied:
(or is tmpdir something else .... some py.path?)

orbeckst replied:
(And yes .... you took the old code, it's just possible that this code is so old that it predates the ability of readers/writers to work with Path objects.)

pbuslaev replied:
After double checking, I think that TRRFile and XTCFile require str format

pbuslaev replied:
So I will push this back for now

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_xdr.py:
```diff
@@ -245,6 +278,31 @@ def test_Writer(self, tmpdir):
         )
 
+    def test_Writer_with_dt_setting(self, tmpdir):
+        universe = mda.Universe(GRO, self.filename, convert_units=True)
```
I don't like replicating the code of `test_Writer`, less copy&paste makes for more maintainable code. Can you please refactor so that the two methods share most of their code? Perhaps there's even a way to make it a parametrized test.

orbeckst reviewed (changes requested):
Overall this is looking pretty solid. Please see comments below. Could you please address them?

orbeckst committed:
Merge branch 'develop' into develop

orbeckst commented:
@pbuslaev please resolve conflicts. Ping when you've addressed comments and then I'll have a look again. Please don't resolve comments — I'll do that when I review again. Thanks.

pbuslaev committed:
Fixing comments from code review

pbuslaev commented:
@orbeckst I think all should be addressed now.

pbuslaev committed:
Merge branch 'develop' into develop

pbuslaev committed:
Revert outfile to str in Writer test

orbeckst commented on package/CHANGELOG:
```diff
@@ -59,6 +59,7 @@ Enhancements
  * Add check and warning for empty (all zero) coordinates in RDKit converter
    (PR #4824)
  * Added `precision` for XYZWriter (Issue #4775, PR #4771) 
+ * Added timestep support for XDR writers and readers(Issue #4905, PR #4908)
```
space before (

orbeckst replied:
IMPORTANT: move this (together with your GitHub handle) to the 2.10.0 section!!

orbeckst commented on package/MDAnalysis/coordinates/XDR.py:
```diff
@@ -148,6 +149,10 @@ def __init__(
+        dt : float (optional)
+            timestep in MDAnalysis units to load trajectory with
+            If dt is None, the time is taken from the xdr file
+            Else, the time is set to dt * frame
```
Please check that this renders correctly in the docs. Please also use markup as follows for arguments and for Python names:
```reST
If `dt` is ``None`` ... else the time is set to `dt` * frame
```

orbeckst commented on package/MDAnalysis/coordinates/XDR.py:
```diff
@@ -327,13 +337,18 @@ def __init__(self, filename, n_atoms, convert_units=True, **kwargs):
             convert from MDAnalysis units to format specific units
+        dt : float (optional)
+            timestep in MDAnalysis units to write trajectory with
+            If dt is None, time for a frame is set from the timestep
```
same comment as above on formatting

orbeckst commented on package/MDAnalysis/coordinates/XTC.py:
```diff
@@ -64,10 +64,16 @@ def __init__(self, filename, n_atoms, convert_units=True,
             convert into MDAnalysis units
+        dt : float (optional)
+            timestep in MDAnalysis units to load trajectory with
+            If dt is None, time for a frame is set from the timestep
```
same comment on formatting

orbeckst replied:
Please check that you use "to load trajectory with" for **readers** and "to write trajectory with" for **writers**. I think I saw some instances where this just seemed copied and pasted across.

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_xdr.py:
```diff
@@ -244,6 +278,14 @@ def test_Writer(self, tmpdir):
+                u.trajectory.totaltime,
+                dt,
+                3,
+                err_msg="wrong total length of trajectory upon setting",
```
upon setting dt explicitly

orbeckst reviewed (changes requested):
Your changes overall look good and I only have minor comments.

However, all the tests fail. Please check where your code broke them.

pbuslaev commented:
Hi @orbeckst 

As I mentioned before

> Ok, I now remember why, I had to add this. Without dt option, tests were breaking.

> For instance this [test](https://github.com/pbuslaev/mdanalysis/blob/f4f5e7d3532e77dcde73f4241ec4ea3f90a10dbe/testsuite/MDAnalysisTests/coordinates/test_dcd.py#L300) starts failing, because dt is set for the writer. So there are two choices, as I see it:

> Leave this option and document it
> Modify failing tests
>(?) maybe there is some more complex logics that can be implemented
> @orbeckst I can take either approach, but I think it is up to you to decide what is better.

So here I am waiting for your input

orbeckst committed:
remove xtc from test_dcd.py::test_writer_dt

- removed XTC test in the DCD tests for setting dt in the writer
- ensure that dt is actually only changed when writing

orbeckst commented:
I changed the test. These tests are old code and it makes little sense that we are testing XTC together with DCD. 

Furthermore, the test did not even test the XTCWriter properly: we had set dt on reading, presumably because that was the only way to make it pass for the XTCWriter, but then setting dt on writing did not do anything else. Only with your addition is the XTCWriter now able to use dt, and you test this functionality elsewhere.

orbeckst committed:
update TestTRRWriter.test_data_preservation

Do not explicitly set dt for the Writer. (In the past, this did not
do anything for TRR but with this new change it forces the
recalculation of times and that introduces inaccuracies at rtol=1e-6. The test does not really check dt so no need to set it.

orbeckst commented:
Let's see if these changes to the tests make the tests more robust. I note that we seem to have been assuming in the past that setting dt for XTC and TRR worked, even though it did not.

orbeckst commented:
@pbuslaev I think the tests are now passing. Could you please address the small doc issues and then it looks ready for merging.

orbeckst commented:
(I enabled auto-merge, i.e., the PR will get merged as soon as tests pass and the PR is approved.)

pbuslaev committed:
Updated doc strings

pbuslaev committed:
Merge branch 'develop' into develop

pbuslaev committed:
Updated docstring style

pbuslaev commented:
@orbeckst I think I have addressed all the comments. The tests are passing and docs are rendered correctly.

orbeckst committed:
Update CHANGELOG

orbeckst reviewed (approved):
Thanks for addressing all my comments.

Once the CI passes again, we can merge.

Thank you for your first contribution @pbuslaev !

pbuslaev commented:
Thank you for your help and review @orbeckst

orbeckst committed:
Merge branch 'develop' into develop

orbeckst commented:
Hooray, your first PR is merged 🎉 ! Thank you for all your work. I hope we see more contributions from you in the future.
